pkg/agent/multicluster/pod_route_controller.go

```diff
@@ -0,0 +1,393 @@
+) *MCPodRouteController {
+	controller := &MCPodRouteController{
+		k8sClient:       k8sClient,
+		mcClient:        mcClient,
+		ovsBridgeClient: ovsBridgeClient,
+		ofClient:        client,
+		interfaceStore:  interfaceStore,
+		nodeConfig:      nodeConfig,
+		podQueue:        workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "MCPodRouteControllerForPod"),
+		gwQueue:         workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "MCPodRouteControllerForGateway"),
+		gwInformer:      gwInformer.Informer(),
+		gwLister:        gwInformer.Lister(),
+		podWorkerStopCh: make(chan struct{}),
+	}
+
+	controller.gwInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				controller.enqueueGateway(cur)
+			},
+			// Gateway UPDATE event doesn't impact Pod flows, so ignore it.
+			DeleteFunc: func(old interface{}) {
+				controller.enqueueGateway(old)
+			},
+		},
+		resyncPeriod,
+	)
+	return controller
+}
+
+func podIPIndexFunc(obj interface{}) ([]string, error) {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		return nil, fmt.Errorf("obj is not Pod: %+v", obj)
+	}
+	if isValidPod(pod) {
+		return []string{pod.Status.PodIP}, nil
+	}
+	return []string{}, nil
+}
+
+func (c *MCPodRouteController) createPodInformer() {
+	listOptions := func(options *metav1.ListOptions) {
+		options.FieldSelector = fields.OneTermNotEqualSelector("spec.nodeName", c.nodeConfig.Name).String()
+	}
+	c.podInformer = coreinformers.NewFilteredPodInformer(
+		c.k8sClient,
+		metav1.NamespaceAll,
+		0,
+		cache.Indexers{podIndexKey: podIPIndexFunc},
+		listOptions,
+	)
+	c.podInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				c.createPod(cur)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				c.updatePod(old, cur)
+			},
+			DeleteFunc: func(old interface{}) {
+				c.deletePod(old)
+			},
+		},
+		resyncPeriod,
+	)
+	c.podLister = corelisters.NewPodLister(c.podInformer.GetIndexer())
+}
+
+func (c *MCPodRouteController) enqueueGateway(obj interface{}) {
+	_, isGW := obj.(*mcv1alpha1.Gateway)
+	if !isGW {
+		deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.ErrorS(nil, "Received unexpected object", "object", obj)
+			return
+		}
+		_, ok = deletedState.Obj.(*mcv1alpha1.Gateway)
+		if !ok {
+			klog.ErrorS(nil, "DeletedFinalStateUnknown contains non-Gateway object", "object", deletedState.Obj)
+			return
+		}
+	}
+	c.gwQueue.Add(dummyKey)
+}
+
+func (c *MCPodRouteController) createPod(obj interface{}) {
+	pod := obj.(*corev1.Pod)
+	if !isValidPod(pod) {
+		return
+	}
+	c.podQueue.Add(pod.Status.PodIP)
+}
+
+func (c *MCPodRouteController) updatePod(old, cur interface{}) {
+	oldPod := old.(*corev1.Pod)
+	curPod := cur.(*corev1.Pod)
+
+	if !isValidPod(curPod) && !isValidPod(oldPod) {
```

Comment:
Define two booleans to avoid func calls multiple times.

Reply:
Done